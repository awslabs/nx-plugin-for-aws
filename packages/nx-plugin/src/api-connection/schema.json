{
  "$schema": "https://json-schema.org/schema",
  "$id": "APIConnection",
  "title": "",
  "type": "object",
  "properties": {
    "sourceProject": {
      "type": "string",
      "description": "The source project which will call the API",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Select the source project which will call the API",
      "x-dropdown": "projects"
    },
    "targetProject": {
      "type": "string",
      "description": "The target project containing your API",
      "x-prompt": "Select the target project containing your API",
      "x-dropdown": "projects"
    }
  },
  "required": ["sourceProject", "targetProject"]
}
