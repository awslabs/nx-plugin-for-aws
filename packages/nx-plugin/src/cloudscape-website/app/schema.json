{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    }
  ],
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z][^:]*$"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "default": "packages",
      "x-prompt": "What directory would you like to store your application in?"
    }
  },
  "required": ["name"],
  "examplesFile": "../../../docs/application-examples.md"
}
