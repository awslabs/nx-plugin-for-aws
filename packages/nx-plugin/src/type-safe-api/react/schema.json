{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "type-safe-api-react",
  "title": "Integrate a Type Safe API with a React website",
  "type": "object",
  "properties": {
    "hooksLibraryProject": {
      "type": "string",
      "description": "The name of the Type Safe API generated react hooks library project",
      "x-prompt": "The react hooks library project for the Type Safe API to integrate",
      "x-dropdown": "projects"
    },
    "frontendProject": {
      "type": "string",
      "description": "The name of the React project to add the Type Safe API integration to",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "React application's project name.",
      "x-dropdown": "projects"
    },
    "auth": {
      "type": "string",
      "description": "",
      "x-prompt": "Select your API authentication strategy",
      "x-dropdown": ["IAM", "None"],
      "default": "IAM"
    }
  },
  "required": ["hooksLibraryProject", "frontendProject", "auth"]
}
