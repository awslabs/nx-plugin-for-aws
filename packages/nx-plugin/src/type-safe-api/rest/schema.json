{
  "$schema": "https://json-schema.org/schema",
  "$id": "TypeSafeApiModel",
  "title": "Create a Type Safe REST API",
  "description": "Create a Type Safe REST API",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the API.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What would you like your API to be called? e.g. MyApi",
      "x-priority": "important"
    },
    "modelLanguage": {
      "description": "The language you would like to model your api in.",
      "type": "string",
      "enum": ["openapi", "typespec"],
      "default": "typespec",
      "x-prompt": "What language would you like to use to model your API?",
      "x-priority": "important"
    },
    "infrastructureLanguage": {
      "description": "The language you would like to write your infrastructure in.",
      "type": "string",
      "enum": ["typescript"],
      "default": "typescript",
      "x-prompt": "What language are you using to write CDK infrastructure?",
      "x-priority": "important"
    },
    "handlerLanguages": {
      "description": "The languages you would like to write lambda handlers in.",
      "type": "array",
      "items": {
        "type": "string",
        "enum": ["typescript"]
      },
      "default": ["typescript"],
      "x-prompt": "What languages (if any) would you like to write lambda handlers in?",
      "x-priority": "important"
    },
    "libraries": {
      "description": "The libraries to generate from your API",
      "type": "array",
      "items": {
        "type": "string",
        "enum": ["typescript-react-hooks"]
      },
      "default": ["typescript-react-hooks"],
      "x-prompt": "What libraries would you like to generate?",
      "x-priority": "important"
    },
    "runtimeLanguages": {
      "description": "The languages you would like to write client or server code in, in addition to the infrastructure or handler languages.",
      "type": "array",
      "items": {
        "type": "string",
        "enum": ["typescript"]
      },
      "default": [],
      "x-prompt": "What other languages (if any) would you like to write client or server code in?"
    },
    "directory": {
      "type": "string",
      "description": "Parent directory where the api model is placed.",
      "default": "packages",
      "x-prompt": "Which directory do you want to create the project in?"
    },
    "scope": {
      "type": "string",
      "description": "Scope for API packages, e.g. @my-company. If omitted this will be inferred",
      "pattern": "^@[a-z0-9-]+(?:/[a-z0-9-]+)?$",
      "x-prompt": "What namespace would you like your library to have? i.e: @myapi"
    },
    "subDirectory": {
      "type": "string",
      "description": "The sub directory the API is placed in. By default this is the API name.",
      "x-prompt": "Which sub directory do you want to create the API in? (By default this is the API name)"
    }
  },
  "required": ["name", "modelLanguage", "infrastructureLanguage"]
}