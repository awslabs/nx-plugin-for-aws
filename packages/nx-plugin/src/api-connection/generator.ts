/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { joinPathFragments, ProjectConfiguration, Tree } from '@nx/devkit';
import { ApiConnectionGeneratorSchema } from './schema';
import trpcReactGenerator from '../trpc/react/generator';
import { hasExportDeclaration } from '../utils/ast';
import { readToml } from '../utils/toml';
import fastApiReactGenerator from '../py/fast-api/react/generator';
import { readProjectConfigurationUnqualified } from '../utils/nx';

/**
 * List of supported source and target project types for api connections
 */
const SUPPORTED_PROJECT_TYPES = [
  'ts#trpc-api',
  'py#fast-api',
  'react',
] as const;

type ProjectType = (typeof SUPPORTED_PROJECT_TYPES)[number];

type Connection = { source: ProjectType; target: ProjectType };

/**
 * Enumerates the supported project connections
 */
const SUPPORTED_CONNECTIONS = [
  { source: 'react', target: 'ts#trpc-api' },
  { source: 'react', target: 'py#fast-api' },
] satisfies Connection[];

type ConnectionKey = (typeof SUPPORTED_CONNECTIONS)[number] extends infer C
  ? C extends Connection
    ? `${C['source']} -> ${C['target']}`
    : never
  : never;

/**
 * Generators for each connection type
 */
const CONNECTION_GENERATORS = {
  'react -> ts#trpc-api': (tree, options) =>
    trpcReactGenerator(tree, {
      frontendProjectName: options.sourceProject,
      backendProjectName: options.targetProject,
    }),
  'react -> py#fast-api': (tree, options) =>
    fastApiReactGenerator(tree, {
      frontendProjectName: options.sourceProject,
      fastApiProjectName: options.targetProject,
    }),
} satisfies Record<
  ConnectionKey,
  (tree: Tree, options: ApiConnectionGeneratorSchema) => Promise<any>
>;

/**
 * Generator for a connection from a project to an API project
 */
export const apiConnectionGenerator = async (
  tree: Tree,
  options: ApiConnectionGeneratorSchema,
) => {
  const sourceType = determineProjectType(tree, options.sourceProject);
  const targetType = determineProjectType(tree, options.targetProject);

  if (!sourceType || !SUPPORTED_PROJECT_TYPES.includes(sourceType)) {
    throw new Error(
      `This generator does not support selected source project ${options.sourceProject}`,
    );
  }
  if (!targetType || !SUPPORTED_PROJECT_TYPES.includes(targetType)) {
    throw new Error(
      `This generator does not support selected target project ${options.targetProject}`,
    );
  }

  const connection = SUPPORTED_CONNECTIONS.find(
    (c) => c.source === sourceType && c.target === targetType,
  );

  if (!connection) {
    throw new Error(
      `This generator does not support a connection from ${options.sourceProject} (${sourceType}) to ${options.targetProject} (${targetType})`,
    );
  }

  return await CONNECTION_GENERATORS[
    `${connection.source} -> ${connection.target}`
  ](tree, options);
};

/**
 * Determine whether the given project is of a known project type
 */
export const determineProjectType = (
  tree: Tree,
  projectName: string,
): ProjectType | undefined => {
  const projectConfiguration = readProjectConfigurationUnqualified(
    tree,
    projectName,
  );

  // NB: if adding new checks, ensure these go from most to least specific
  // eg. react website is more specific than typescript project

  if (isTrpcApi(tree, projectConfiguration)) {
    return 'ts#trpc-api';
  }

  if (isFastApi(tree, projectConfiguration)) {
    return 'py#fast-api';
  }

  if (isReact(tree, projectConfiguration)) {
    return 'react';
  }

  return undefined;
};

const sourceRoot = (projectConfiguration: ProjectConfiguration) =>
  projectConfiguration.sourceRoot ??
  joinPathFragments(projectConfiguration.root, 'src');

const isTrpcApi = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // If the project.json says it's a trpc api, there's no need for introspection
  if ((projectConfiguration.metadata as any)?.apiType === 'trpc') {
    return true;
  }

  // Find the src/index.ts
  const indexTs = tree.read(
    joinPathFragments(sourceRoot(projectConfiguration), 'index.ts'),
    'utf-8',
  );
  if (indexTs === null) {
    return false;
  }

  // If the index file exports an AppRouter, it's a trpc api
  if (hasExportDeclaration(indexTs, 'AppRouter')) {
    return true;
  }

  // If there's a src/router.ts or src/lambdas/router.ts which exports an AppRouter, it's a trpc api
  const trpcRouter =
    tree.read(
      joinPathFragments(sourceRoot(projectConfiguration), 'router.ts'),
      'utf-8',
    ) ??
    tree.read(
      joinPathFragments(
        sourceRoot(projectConfiguration),
        'lambdas',
        'router.ts',
      ),
      'utf-8',
    );

  if (trpcRouter === null) {
    return false;
  }

  if (hasExportDeclaration(trpcRouter, 'AppRouter')) {
    return true;
  }

  return false;
};

const isReact = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // if there's a main.tsx, it's a react app
  return tree.exists(
    joinPathFragments(sourceRoot(projectConfiguration), 'main.tsx'),
  );
};

const isFastApi = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // If the project.json says it's a fast api, there's no need for introspection
  if ((projectConfiguration.metadata as any)?.apiType === 'fast-api') {
    return true;
  }

  const pyProjectPath = joinPathFragments(
    projectConfiguration.root,
    'pyproject.toml',
  );
  if (tree.exists(pyProjectPath)) {
    const pyProject = readToml(tree, pyProjectPath);
    if (((pyProject as any)?.project?.dependencies ?? []).includes('fastapi')) {
      return true;
    }
  }

  return false;
};

export default apiConnectionGenerator;
