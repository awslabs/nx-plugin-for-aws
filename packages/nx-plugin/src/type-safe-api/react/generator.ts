/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  Tree,
  GeneratorCallback,
  readProjectConfiguration,
  installPackagesTask,
  joinPathFragments,
  generateFiles,
  addDependenciesToPackageJson,
} from '@nx/devkit';
import { TypeSafeApiReactGeneratorSchema } from './schema';
import runtimeConfigGenerator from '../../cloudscape-website/runtime-config/generator';
import kebabCase from 'lodash.kebabcase';
import camelCase from 'lodash.camelcase';
import upperFirst from 'lodash.upperfirst';
import { addDefaultImportDeclarations, addJsxComponentWrapper } from '../../utils/typescript-ast';
import { toScopeAlias } from '../../utils/npm-scope';
import { withVersions } from '../../utils/versions';
import { formatFilesInSubtree } from '../../utils/format';

export async function typeSafeApiReactGenerator(
  tree: Tree,
  options: TypeSafeApiReactGeneratorSchema
): Promise<GeneratorCallback> {
  const frontendProject = readProjectConfiguration(tree, options.frontendProject);
  const hooksLibraryProject = readProjectConfiguration(tree, options.hooksLibraryProject);

  // Add runtime config provider to the app if not already present
  await runtimeConfigGenerator(tree, {
    project: options.frontendProject,
  });

  const apiName = (hooksLibraryProject.metadata as any)?.apiName;
  if (!apiName) {
    throw new Error(`${options.hooksLibraryProject} does not appear to be a Type Safe API generated hooks project`);
  }

  const apiNameKebabCase = kebabCase(apiName);
  const apiNamePascalCase = upperFirst(camelCase(apiName));

  generateFiles(tree, joinPathFragments(__dirname, 'files'), frontendProject.root, {
    apiName,
    apiNameKebabCase,
    apiNamePascalCase,
    generatedHooksLibraryPackage: toScopeAlias(options.hooksLibraryProject),
    auth: options.auth,
  });

  const mainTsxPath = joinPathFragments(
    frontendProject.sourceRoot,
    'main.tsx'
  );

  if (!tree.exists(mainTsxPath)) {
    throw new Error(
      `Could not find main.tsx in ${frontendProject.sourceRoot}`
    );
  }

  const mainTsxContents = tree.read(mainTsxPath).toString();

  const hooksProviderComponentName = `${apiNamePascalCase}HooksProvider`;

  const updatedImports = addDefaultImportDeclarations(mainTsxContents, [
    { import: hooksProviderComponentName, from: `./components/${hooksProviderComponentName}` },
  ]);

  const mainTsxUpdatedContents = addJsxComponentWrapper(updatedImports, 'App', hooksProviderComponentName);

  if (!mainTsxUpdatedContents) {
    throw new Error('Could not locate App component in main.tsx');
  }

  if (mainTsxContents !== mainTsxUpdatedContents) {
    tree.write(mainTsxPath, mainTsxUpdatedContents);
  }

  addDependenciesToPackageJson(
    tree,
    withVersions([
      '@tanstack/react-query',
      '@aws-northstar/ui',
    ]),
    {}
  );

  await formatFilesInSubtree(tree, frontendProject.root);

  return () => {
    installPackagesTask(tree);
  };
}

export default typeSafeApiReactGenerator;
