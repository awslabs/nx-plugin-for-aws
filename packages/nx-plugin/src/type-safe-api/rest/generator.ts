/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  Tree,
} from '@nx/devkit';
import {
  InferredTypeSafeRestApiSchema,
  TypeSafeRestApiGeneratorSchema,
} from './schema';
import kebabCase from 'lodash.kebabcase';
import camelCase from 'lodash.camelcase';
import upperFirst from 'lodash.upperfirst';
import { join } from 'path';
import { typeSpecRestModelGenerator } from './model/typespec';
import { openApiRestModelGenerator } from './model/openapi';
import { getNpmScopePrefix, toScopeAlias } from '../../utils/npm-scope';
import { typeScriptRestRuntimeGenerator } from './runtime/typescript';
import { uniq } from '../../utils/string';
import { typeScriptRestInfrastructureGenerator } from './infrastructure/typescript';
import { typeScriptRestHandlersGenerator } from './handlers/typescript';
import { typeScriptRestHooksLibraryGenerator } from './library/hooks';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  sharedConstructsGenerator,
} from '../../utils/shared-constructs';
import { withVersions } from '../../utils/versions';
import { writeExportDeclarations } from '../../utils/typescript-ast';

/**
 * Generates a Type Safe REST API
 */
export const typeSafeRestApiGenerator = async (
  tree: Tree,
  schema: TypeSafeRestApiGeneratorSchema
): Promise<GeneratorCallback> => {
  const nameKebabCase = kebabCase(schema.name);
  const scope = schema.scope ? `${schema.scope}/` : getNpmScopePrefix(tree);
  const fullyQualifiedApiName = `${scope}${nameKebabCase}`;
  const namePascalCase = upperFirst(camelCase(schema.name));
  const dir = join(
    schema.directory ?? '.',
    schema.subDirectory ?? nameKebabCase
  );

  const runtimeLanguages = uniq([
    schema.infrastructureLanguage,
    ...schema.handlerLanguages,
    ...schema.runtimeLanguages,
  ]);

  const options: InferredTypeSafeRestApiSchema = {
    ...schema,
    fullyQualifiedApiName,
    namePascalCase,
    nameKebabCase,
    dir,
    runtimeLanguages,
    model: {
      fullyQualifiedName: `${fullyQualifiedApiName}-model`,
      dir: join(dir, 'model'),
      outputSpecPath: join('dist', dir, 'model', 'prepared', 'api.json'),
    },
    runtime: {
      dir: join(dir, 'generated', 'runtime'),
      typescript: runtimeLanguages.includes('typescript')
        ? {
            dir: join(dir, 'generated', 'runtime', 'typescript'),
            fullyQualifiedName: `${fullyQualifiedApiName}-runtime-typescript`,
          }
        : undefined,
    },
    infrastructure: {
      dir: join(dir, 'generated', 'infrastructure'),
      typescript:
        schema.infrastructureLanguage === 'typescript'
          ? {
              dir: join(dir, 'generated', 'infrastructure', 'typescript'),
              fullyQualifiedName: `${fullyQualifiedApiName}-infrastructure-typescript`,
            }
          : undefined,
    },
    handlers: {
      dir: join(dir, 'handlers'),
      typescript: schema.handlerLanguages.includes('typescript')
        ? {
            dir: join(dir, 'handlers', 'typescript'),
            fullyQualifiedName: `${fullyQualifiedApiName}-handlers-typescript`,
            assetPath: join('dist', dir, 'handlers', 'typescript-lambda'),
          }
        : undefined,
    },
    library: {
      dir: join(dir, 'generated', 'libraries'),
      typescriptHooks: schema.libraries.includes('typescript-react-hooks')
        ? {
            dir: join(dir, 'generated', 'libraries', 'typescript-react-hooks'),
            fullyQualifiedName: `${fullyQualifiedApiName}-typescript-react-hooks`,
          }
        : undefined,
    },
  };

  // Generate the model if needed
  if (!tree.exists(options.model.dir)) {
    switch (schema.modelLanguage) {
      case 'typespec': {
        await typeSpecRestModelGenerator(tree, options);
        break;
      }
      case 'openapi': {
        await openApiRestModelGenerator(tree, options);
        break;
      }
      default: {
        throw new Error(
          `Model language ${schema.modelLanguage} is not supported!`
        );
      }
    }
  }

  if (!tree.exists(options.infrastructure.dir)) {
    switch (options.infrastructureLanguage) {
      case 'typescript': {
        await typeScriptRestInfrastructureGenerator(tree, options);
        break;
      }
      default: {
        throw new Error(
          `Infrastructure language ${options.infrastructureLanguage} is not supported!`
        );
      }
    }
  }

  for (const handlerLanguage of options.handlerLanguages) {
    switch (handlerLanguage) {
      case 'typescript': {
        if (!tree.exists(options.handlers.typescript!.dir)) {
          await typeScriptRestHandlersGenerator(tree, options);
        }
        break;
      }
      default: {
        throw new Error(
          `Handler language ${handlerLanguage} is not supported!`
        );
      }
    }
  }

  for (const runtimeLanguage of options.runtimeLanguages) {
    switch (runtimeLanguage) {
      case 'typescript': {
        if (!tree.exists(options.runtime.typescript!.dir)) {
          await typeScriptRestRuntimeGenerator(tree, options);
        }
        break;
      }
      default: {
        throw new Error(
          `Runtime language ${runtimeLanguage} is not supported!`
        );
      }
    }
  }

  for (const library of options.libraries) {
    switch (library) {
      case 'typescript-react-hooks': {
        if (!tree.exists(options.library.typescriptHooks!.dir)) {
          await typeScriptRestHooksLibraryGenerator(tree, options);
        }
        break;
      }
      default: {
        throw new Error(`Library ${library} is not supported!`);
      }
    }
  }

  // Add the shared construct for the api
  if (options.infrastructureLanguage === 'typescript') {
    await addTypeScriptSharedConstructs(tree, options);
  }

  return () => {
    installPackagesTask(tree);
  };
};

const addTypeScriptSharedConstructs = async (
  tree: Tree,
  options: InferredTypeSafeRestApiSchema
) => {
  // Call shared constructs generator for API wrapper construct
  await sharedConstructsGenerator(tree);

  const constructsPath = joinPathFragments(
    PACKAGES_DIR,
    SHARED_CONSTRUCTS_DIR,
    'src',
    options.nameKebabCase,
    'index.ts'
  );

  if (!tree.exists(constructsPath)) {
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', SHARED_CONSTRUCTS_DIR),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR),
      {
        ...options,
        generatedInfrastructurePackage: toScopeAlias(
          options.infrastructure.typescript!.fullyQualifiedName
        ),
      }
    );

    addDependenciesToPackageJson(
      tree,
      withVersions(['constructs', 'aws-cdk-lib']),
      {}
    );

    const sharedConstructsIndexTsPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'index.ts'
    );

    writeExportDeclarations(tree, sharedConstructsIndexTsPath, [
      `./${options.nameKebabCase}/index.js`,
    ]);
  }
};

export default typeSafeRestApiGenerator;
