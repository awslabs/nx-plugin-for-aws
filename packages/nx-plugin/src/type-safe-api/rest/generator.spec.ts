/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree } from '@nx/devkit';

import { typeSafeRestApiGenerator } from './generator';
import { TypeSafeRestApiGeneratorSchema } from './schema';

describe('type-safe-api rest generator', () => {
  let tree: Tree;
  const options: TypeSafeRestApiGeneratorSchema = {
    name: 'test-api',
    modelLanguage: 'typespec',
    infrastructureLanguage: 'typescript',
    handlerLanguages: ['typescript'],
    runtimeLanguages: ['typescript'],
    libraries: ['typescript-react-hooks'],
  };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should generate the overarching project structure', async () => {
    await typeSafeRestApiGenerator(tree, options);

    // Verify model files are generated
    expect(tree.exists('test-api/model')).toBeTruthy();

    // Verify runtime files are generated
    expect(tree.exists('test-api/generated/runtime/typescript')).toBeTruthy();

    // Verify infrastructure files are generated
    expect(tree.exists('test-api/generated/infrastructure/typescript')).toBeTruthy();

    // Verify handler files are generated
    expect(tree.exists('test-api/handlers/typescript')).toBeTruthy();

    // Verify library files are generated
    expect(tree.exists('test-api/generated/libraries/typescript-react-hooks')).toBeTruthy();
  });

  it('should generate a shared construct for deploying the API', async () => {
    await typeSafeRestApiGenerator(tree, options);

    // Check if shared construct files exist
    expect(tree.exists('packages/common/constructs/src/test-api/index.ts')).toBeTruthy();

    // Verify the shared construct content matches snapshots
    expect(
      tree.read('packages/common/constructs/src/test-api/index.ts')?.toString()
    ).toMatchSnapshot();

    // Verify the construct is exported from common constructs index
    const commonConstructsIndex = tree.read('packages/common/constructs/src/index.ts')?.toString();
    expect(commonConstructsIndex).toContain("export * from './test-api/index.js';");
  });

  it('should generate files with OpenAPI model language', async () => {
    const openApiOptions = {
      ...options,
      modelLanguage: 'openapi' as const,
    };

    await typeSafeRestApiGenerator(tree, openApiOptions);

    expect(tree.exists('test-api/model/src/main.yaml')).toBeTruthy();
  });

  it('should generate files with TypeSpec model language', async () => {
    const typeSpecOptions = {
      ...options,
      modelLanguage: 'typespec' as const,
    };

    await typeSafeRestApiGenerator(tree, typeSpecOptions);

    expect(tree.exists('test-api/model/src/main.tsp')).toBeTruthy();
  });

  it('should throw error for unsupported model language', async () => {
    const invalidOptions = {
      ...options,
      modelLanguage: 'invalid' as any,
    };

    await expect(
      async () => await typeSafeRestApiGenerator(tree, invalidOptions)
    ).rejects.toThrow('Model language invalid is not supported!');
  });

  it('should throw error for unsupported runtime language', async () => {
    const invalidOptions = {
      ...options,
      runtimeLanguages: ['invalid' as any],
    };

    await expect(
      async () => await typeSafeRestApiGenerator(tree, invalidOptions)
    ).rejects.toThrow('Runtime language invalid is not supported!');
  });

  it('should throw error for unsupported handler language', async () => {
    const invalidOptions = {
      ...options,
      handlerLanguages: ['invalid' as any],
    };

    await expect(
      async () => await typeSafeRestApiGenerator(tree, invalidOptions)
    ).rejects.toThrow('Handler language invalid is not supported!');
  });

  it('should throw error for unsupported infrastructure language', async () => {
    const invalidOptions = {
      ...options,
      infrastructureLanguage: 'invalid' as any,
    };

    await expect(
      async () => await typeSafeRestApiGenerator(tree, invalidOptions)
    ).rejects.toThrow('Infrastructure language invalid is not supported!');
  });

  it('should throw error for unsupported library', async () => {
    const invalidOptions = {
      ...options,
      libraries: ['invalid' as any],
    };

    await expect(
      async () => await typeSafeRestApiGenerator(tree, invalidOptions)
    ).rejects.toThrow('Library invalid is not supported!');
  });

  it('should use custom directory when provided', async () => {
    const optionsWithDir = {
      ...options,
      directory: 'custom-dir',
    };

    await typeSafeRestApiGenerator(tree, optionsWithDir);

    expect(tree.exists('custom-dir/test-api/model')).toBeTruthy();
  });

  it('should use custom subdirectory when provided', async () => {
    const optionsWithSubDir = {
      ...options,
      subDirectory: 'sub-dir',
    };

    await typeSafeRestApiGenerator(tree, optionsWithSubDir);

    expect(tree.exists('sub-dir/model')).toBeTruthy();
  });

  it('should use scope when provided', async () => {
    const optionsWithScope = {
      ...options,
      scope: '@test-scope',
    };

    await typeSafeRestApiGenerator(tree, optionsWithScope);

    // Verify files are generated with scoped package names
    expect(tree.exists('test-api/model')).toBeTruthy();

    // Verify project.json contains scoped package names
    const projectJson = JSON.parse(tree.read('test-api/generated/runtime/typescript/project.json').toString());
    expect(projectJson.name).toBe('@test-scope/test-api-runtime-typescript');

    // Verify infrastructure project uses scoped name
    const infrastructureProjectJson = JSON.parse(tree.read('test-api/generated/infrastructure/typescript/project.json').toString());
    expect(infrastructureProjectJson.name).toBe('@test-scope/test-api-infrastructure-typescript');

    // Verify handlers project uses scoped name
    const handlersProjectJson = JSON.parse(tree.read('test-api/handlers/typescript/project.json').toString());
    expect(handlersProjectJson.name).toBe('@test-scope/test-api-handlers-typescript');

    // Verify library project uses scoped name
    const libraryProjectJson = JSON.parse(tree.read('test-api/generated/libraries/typescript-react-hooks/project.json').toString());
    expect(libraryProjectJson.name).toBe('@test-scope/test-api-typescript-react-hooks');

    // Verify shared constructs import uses scoped package name
    const sharedConstructsContent = tree.read('packages/common/constructs/src/test-api/index.ts').toString();
    expect(sharedConstructsContent).toContain("from ':test-scope/test-api-infrastructure-typescript'");
  });

  it('should add required dependencies', async () => {
    await typeSafeRestApiGenerator(tree, options);

    const packageJson = JSON.parse(tree.read('package.json').toString());

    expect(packageJson.dependencies).toMatchObject({
      'constructs': expect.any(String),
      'aws-cdk-lib': expect.any(String),
    });
  });

  it('should not regenerate existing directories', async () => {
    // First run to create files
    await typeSafeRestApiGenerator(tree, options);

    // Create a marker file to detect if regeneration occurs
    tree.write('test-api/model/marker.txt', 'original content');

    // Run generator again
    await typeSafeRestApiGenerator(tree, options);

    // Verify marker file still exists with original content
    expect(tree.read('test-api/model/marker.txt').toString()).toBe('original content');
  });
});
