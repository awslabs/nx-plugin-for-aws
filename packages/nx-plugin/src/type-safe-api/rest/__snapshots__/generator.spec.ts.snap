// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`type-safe-api rest generator > should generate a shared construct for deploying the API 1`] = `
"import { Construct } from 'constructs';
import { Effect, IRole, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { RuntimeConfig } from '../runtime-config/index.js';
import { Api as ApiConstruct, ApiProps as ApiConstructProps } from ':proj/test-api-infrastructure-typescript';

export interface TestApiProps extends ApiConstructProps {

}

export class TestApi extends ApiConstruct {
  constructor(scope: Construct, id: string, props: TestApiProps) {
    super(scope, id, props);

    RuntimeConfig.ensure(this).config.typeSafeApis = {
      'test-api': this.api.url!
    };
  }

  public grantInvokeAccess(role: IRole) {
    role.addToPrincipalPolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['execute-api:Invoke'],
        resources: [this.api.arnForExecuteApi('*', '/*', '*')],
      })
    );
  }
}
"
`;
